patch/patcher.go

```diff
@@ -29,22 +30,25 @@ import (
 // - go_enum_name (Enum option) overrides the name of an enum type.
 // - go_value_name (EnumValue option) overrides the name of an enum const.
 type Patcher struct {
-	gen            *protogen.Plugin
-	fset           *token.FileSet
-	files          []*ast.File
-	filesByName    map[string]*ast.File
-	info           *types.Info
-	packages       []*Package
-	packagesByPath map[string]*Package
-	packagesByName map[string]*Package
-	renames        map[protogen.GoIdent]string
-	typeRenames    map[protogen.GoIdent]string
-	valueRenames   map[protogen.GoIdent]string
-	fieldRenames   map[protogen.GoIdent]string
-	methodRenames  map[protogen.GoIdent]string
-	objectRenames  map[types.Object]string
-	tags           map[protogen.GoIdent]string
-	fieldTags      map[types.Object]string
+	gen               *protogen.Plugin
+	fset              *token.FileSet
+	files             []*ast.File
+	filesByName       map[string]*ast.File
+	info              *types.Info
+	packages          []*Package
+	packagesByPath    map[string]*Package
+	packagesByName    map[string]*Package
+	renames           map[protogen.GoIdent]string
+	typeRenames       map[protogen.GoIdent]string
+	valueRenames      map[protogen.GoIdent]string
+	fieldRenames      map[protogen.GoIdent]string
+	methodRenames     map[protogen.GoIdent]string
+	objectRenames     map[types.Object]string
+	tags              map[protogen.GoIdent]string
+	fieldTags         map[types.Object]string
+	fileFieldOptions  *gopb.Options
+	fileValuesOptions *gopb.Options
+	fileInitialisms   map[string]bool
```

Comment:
These are by file, but stored here globally. One file can clobber another file’s data.

I’d prefer to see these fetched on-demand at the point of use, perhaps as a fallback if a given thing doesn’t have more specific options, it can look at the file for global options.

Reply:
Please check the latest version. I have a problem with this last fix. I did fileFieldOptions,fileValuesOptions and fileInitialisms  global because is hard to feetch this data at the point of use. In functions like scanEnum we don't have access to protogen.File. We would have to pass protogen.File (or fileFieldOptions etc). to the functions: scanEnum, scanMessage, scanExtension and then to scanEnumValue, scanField, etc. 
Do you have a better idea?

Reply:
I think it’s possible to walk up to the parent file from the `protoreflect.Descriptor` interface’s `ParentFile` method.

I’m not sure what the ideal interface (or helper interface is) to get it, but I think it should first examine the thing (say a `protogen.Field`) for options, and then look at its `Desc.ParentFile` for a global option.